### PR TITLE
Release Notes for 3.38.0

### PR DESCRIPTION
**Pull request by Emdee89**
## What Has Changed?

Release Notes for SAP Cloud SDK for Java 3.38.0

## Manual Checks?

- [X] Text adheres to the style guide (`vale docs/`)
  - Every sentence is on its own line
  - Headings use [title capitalization](https://capitalizemytitle.com/style/AP/#) (applies also to the sidebar and title of a document)
  - You followed [naming center](https://www.sapbrandtools.com/naming-center/#/dashboard) guidelines when referring to SAP products (e.g. SAP S/4HANA)
- [X] You checked your spelling and grammar (consider using Grammarly, see `CONTRIBUTING.md`)
- [X] You formatted all changed files with prettier (`npm run prettier`)
- [X] You tested if the documentation still builds (`npm run build`)
- [] You verified all new and changed links still work (changing the `id` or name of a file can break links)
- [X] ~You have updated the [feature matrix](https://github.com/SAP/cloud-sdk/blob/main/docs/components/data/features.js) if you add documentation on a new feature or otherwise required.~

Open: The Javadoc links is missing still

### CODE REVIEW

File: docs/java/release-notes.mdx [hunk 1]
```diff
@@ -49,6 +49,120 @@ We highly recommend regularly updating to the latest SDK version. It will help y
...
+      .row("CC02")
+      .row("CC03")
+      .end();
+  ```
+
+- Add support for `SAMLAssertion` authentication type on Cloud Foundry.
+  Refer the [help page](https://help.sap.com/viewer/cca91383641e40ffbe03bdc78f00f681/Cloud/en-US/d81e1683bd434823abf3ceefc4ff157f.html) for more detail on `SAMLAssertion` authentication type.
+- Introduced a new API to build a `Destination` for connecting to services on Cloud Foundry.
+  The information will be read from the `VCAP_SERVICES` environment variable at runtime.
+  For certain authentication types i,e. `NoAuthentication` and `OAuth2ClientCredentials` this API eliminates the need to manually create a destination in the CF account for connecting to services.
```
Comment (artemkovalyov) [inline docs/java/release-notes.mdx:84]:
```suggestion
- We now support the `SAMLAssertion` authentication type on Cloud Foundry.
  Refer to the [help page](https://help.sap.com/viewer/cca91383641e40ffbe03bdc78f00f681/Cloud/en-US/d81e1683bd434823abf3ceefc4ff157f.html) for more detail on the `SAMLAssertion` authentication type.
- We introduced a new API to build a `Destination` for connecting to services on Cloud Foundry.
  The information will be read from the `VCAP_SERVICES` environment variable at runtime.
  For certain authentication types i.e. `NoAuthentication` and `OAuth2ClientCredentials` this API eliminates the need to manually create a destination in the CF account for connecting to services.
```

File: docs/java/release-notes.mdx [hunk 2]
```diff
@@ -49,6 +49,120 @@ We highly recommend regularly updating to the latest SDK version. It will help y
...
+  `ScpCfServiceDestinationLoader.buildClientCredentialsMapping` creates a mapping for loading a destination and provides JSON paths for all required destination properties.
+
+  ```java
+  final HttpDestination destination =
+        ScpCfServiceDestinationLoader.getDestinationForService(
+          "label",
+          "serviceBindingName",
+          AuthenticationType.OAUTH2_CLIENT_CREDENTIALS,
+          mapping);
+  ```
```
Comment (artemkovalyov) [inline docs/java/release-notes.mdx:106]: Do we also have these examples in the documentation for destinations?

File: docs/java/release-notes.mdx [hunk 3]
```diff
@@ -49,6 +49,120 @@ We highly recommend regularly updating to the latest SDK version. It will help y
...
+
+  For certain services (only workflow for now), we provide further convenience that skips the need to provide the `mapping` explicitly.
+  Example:
+
+  ```java
+  final HttpDestination destination =
+        ScpCfServiceDestinationLoader.getDestinationForService(
+          ScpCfServiceDestinationLoader.CfServices.WORKFLOW,
+          "my-workflow");
```
Comment (artemkovalyov) [inline docs/java/release-notes.mdx:115]: Let's add this to the Workflow docs.

File: docs/java/release-notes.mdx [hunk 4]
```diff
@@ -49,6 +49,120 @@ We highly recommend regularly updating to the latest SDK version. It will help y
...
+  ```
+
+- Introduce a new API to construct a `ScpCfDestinationLoader` that allows to override the default timeout for the Time Limiter resilience pattern.
+  If not specified, requests will automatically timeout after 6 seconds.
+
+- Add support for recursive tables and structures in parameters of `BapiRequest` API for both protocols: _JCo_ and _SOAP_. Example:
```
Comment (artemkovalyov) [inline docs/java/release-notes.mdx:121]:
```suggestion
- We introduced a new API to construct a `ScpCfDestinationLoader` that allows overriding the default timeout for the Time Limiter resilience pattern.
  If not specified, requests will automatically timeout after 6 seconds.

- We added support for recursive tables and structures in parameters of `BapiRequest` API for both protocols: _JCo_ and _SOAP_. Example:
```

File: docs/java/release-notes.mdx [hunk 5]
```diff
@@ -49,6 +49,120 @@ We highly recommend regularly updating to the latest SDK version. It will help y
...
+  +       .field("NESTED_FIELD1","STRING","john")
+  +       .field("NESTED_FIELD2","STRING","doe")
+  +       .end() // End nested table
+       // START nested structure
+       .fields("NESTED_STRUCTURE", "NESTED_STRUCTURE_TYPE")
+           .field("latitude", "Axis", "52")
+           .field("longitude", "Axis", "13")
+           .end() // End nested structure
+       .end();
+  ```
```
Comment (artemkovalyov) [inline docs/java/release-notes.mdx:139]: Another reason for a quick example on the BAPI/RFC docs if not yet there.

File: docs/java/release-notes.mdx [hunk 6]
```diff
@@ -49,6 +49,120 @@ We highly recommend regularly updating to the latest SDK version. It will help y
...
+      - Update Resilience4J from `1.6.1` to `1.7.0`
+      - Update Java JWT Java from `3.12.0` to `3.12.1`
+      - Update PMD from `6.30.0` to `6.31.0`
+      - Update Lombok from `1.18.16` to `1.18.18`
+      - Update Jetty from `9.4.35.v20201120` to `9.4.36.v20210114`
+
+### Fixed Issues
+
+- Fix an issue with OData V4 update requests where the serialization of primitives with types like Integer, Boolean and Double leads to null values.
+- Fix an issue with JCo serialization of BAPI requests where adding field parameter values of type `ErpType` led to an exception.
```
Comment (artemkovalyov) [inline docs/java/release-notes.mdx:164]:
```suggestion
- We fixed an issue with OData V4 update requests where the serialization of primitives with types like Integer, Boolean, and Double leads to null values.
- We fixed an issue with JCo serialization of BAPI requests, where adding field parameter values of type `ErpType` led to an exception.
```

File: docs/java/release-notes.mdx [hunk 7]
```diff
@@ -49,6 +49,120 @@ We highly recommend regularly updating to the latest SDK version. It will help y
 ### Fixed issues
 -->
 
+## 3.38.0 - February 11, 2021
+
+[Maven Central](https://search.maven.org/search?q=g:com.sap.cloud.sdk*%20AND%20v:3.38.0) | [Javadoc](still-missing)
```
Comment (artemkovalyov) [inline docs/java/release-notes.mdx:54]: The broken JavaDoc link will fail the pipeline. If it's not available before merge - we can use the link from the last release we have it for.

**Review by artemkovalyov — CHANGES_REQUESTED** [review]
Great release notes!
I suggest we have to update docs with code-snippets we have here if hasn't been done yet.
They must be discoverable by new users who'd skip the old release notes when starting with some SDK features.

File: docs/java/release-notes.mdx [hunk 8]
```diff
@@ -49,6 +49,120 @@ We highly recommend regularly updating to the latest SDK version. It will help y
...
+
+### Compatibility Notes
+
+- The method `getHeaders(DestinationProperties)` of extension API `DestinationHeaderProvider` is deprecated but still functional.
+  Please implement `getHeaders(DestinationRequestContext)` instead.
+  The new argument allows accessing the request specific `URI` as well as the original `DestinationProperties`.
+- The method `ResilienceDecorator::invalidateCache` is deprecated but still functional.
+  Please use `ResilienceDecorator::clearCache` for tenant aware cache invalidation or `ResilienceDecorator::clearAllCacheEntries` to clear all entries for all tenants and principals.
+
```
Comment (newtork) [inline docs/java/release-notes.mdx:63]:
```suggestion
- As part of the update of the OData VDM to SAP S/4HANA Cloud 2102 a few operations were removed. The operations were never supported by the S/4 system (like CRUD support for some entities), so no existing functionality is affected. The following services are affected:
  - _Nota Fiscal – Create, Update_ - `API_LOGBR_NOTAFISCAL_SRV`
  - _Purchase Contracts_ - `API_PURCHASECONTRACT_PROCESS_SRV`
  - _Outbound Delivery (A2X)_ - `API_OUTBOUND_DELIVERY_SRV_0002`
  
```

File: docs/java/release-notes.mdx [hunk 9]
```diff
@@ -49,6 +49,120 @@ We highly recommend regularly updating to the latest SDK version. It will help y
...
+  +       .field("NESTED_FIELD2","STRING","doe")
+  +       .end() // End nested table
+       // START nested structure
+       .fields("NESTED_STRUCTURE", "NESTED_STRUCTURE_TYPE")
+           .field("latitude", "Axis", "52")
+           .field("longitude", "Axis", "13")
+           .end() // End nested structure
+       .end();
+  ```
+
```
Comment (newtork) [inline docs/java/release-notes.mdx:140]:
```suggestion
- Update the OData VDM to the [newest release 2102 of SAP S/4HANA Cloud](https://news.sap.com/2021/02/sap-s4hana-cloud-2102-empowering-businesses-seize-opportunities/). This includes completely new services (available as usual in packages `com.sap.cloud.sdk.s4hana.datamodel.odata.services` and `com.sap.cloud.sdk.s4hana.datamodel.odatav4.services`), new operations in previously existing services, and new entity types. The SDK supports all OData services listed in the [SAP API Business Hub for SAP S/4HANA Cloud](https://api.sap.com/shell/discover/contentpackage/SAPS4HANACloud) with the exception of the following services:
  - (OData V2) _Purchase Requisition_ - `API_PURCHASEREQ_PROCESS_SRV`
  - (OData V2) _Production Order Confirmation_ - `API_PROD_ORDER_CONFIRMATION_2_SRV`
  - (OData V2) _Process Order Confirmation_ - `API_PROC_ORDER_CONFIRMATION_2_SRV`
  - (OData V2) _Attachments_ - `API_CV_ATTACHMENT_SRV`
  - (OData V4) _Contract Accounting: Invoices of Business Partner_ - `API_CABUSPARTINVOICE_0001`
  
```

File: docs/java/release-notes.mdx (hunk 7 above)
Comment (Emdee89) [reply docs/java/release-notes.mdx:54]: I removed it intentionally to avoid forgetting to update the link

File: docs/java/release-notes.mdx (hunk 5 above)
Comment (Emdee89) [reply docs/java/release-notes.mdx:143]: We lack even simpler examples in the BAPI docu, so it's difficult to start there now with such a complex example.
I'll create a backlog item to extend the docu in that realm overall where this example can be covered.

File: docs/java/release-notes.mdx (hunk 3 above)
Comment (Emdee89) [reply docs/java/release-notes.mdx:119]: It's already there: https://sap.github.io/cloud-sdk/docs/java/features/rest/clients/scp-workflow-rest-api#invoke-the-java-client-library

File: docs/java/release-notes.mdx [hunk 10]
```diff
@@ -49,6 +49,130 @@ We highly recommend regularly updating to the latest SDK version. It will help y
 ### Fixed issues
 -->
 
+## 3.38.0 - February 11, 2021
+
+[Maven Central](https://search.maven.org/search?q=g:com.sap.cloud.sdk*%20AND%20v:3.38.0) | [Javadoc](still-missing)
```
Comment (Johannes-Schneider) [inline docs/java/release-notes.mdx:54]:
```suggestion
[Maven Central](https://search.maven.org/search?q=g:com.sap.cloud.sdk*%20AND%20v:3.38.0) | [Javadoc](https://help.sap.com/doc/cfa38ab94e4f481298556d6fdaf3672b/1.0/en-US/index.html)
```

File: docs/java/release-notes.mdx [hunk 11]
```diff
@@ -49,6 +49,130 @@ We highly recommend regularly updating to the latest SDK version. It will help y
...
+
+[Maven Central](https://search.maven.org/search?q=g:com.sap.cloud.sdk*%20AND%20v:3.38.0) | [Javadoc](https://help.sap.com/doc/cfa38ab94e4f481298556d6fdaf3672b/1.0/en-US/index.html)
+
+### Compatibility Notes
+
+- The method `getHeaders(DestinationProperties)` of extension API `DestinationHeaderProvider` is deprecated but still functional.
+  Please implement `getHeaders(DestinationRequestContext)` instead.
+  The new argument allows accessing the request specific `URI` as well as the original `DestinationProperties`.
+- The method `ResilienceDecorator::invalidateCache` is deprecated but still functional.
+  Please use `ResilienceDecorator::clearCache` for tenant aware cache invalidation or `ResilienceDecorator::clearAllCacheEntries` to clear all entries for all tenants and principals.
```
Comment (newtork) [inline docs/java/release-notes.mdx:62]: I'm not sure whether this is good or bad, but i the past we used `#` as type-method delimiter.

```suggestion
- The method `ResilienceDecorator#invalidateCache` is deprecated but still functional.
  Please use `ResilienceDecorator#clearCache` for tenant aware cache invalidation or `ResilienceDecorator#clearAllCacheEntries` to clear all entries for all tenants and principals.
```

File: docs/java/release-notes.mdx [hunk 12]
```diff
@@ -49,6 +49,130 @@ We highly recommend regularly updating to the latest SDK version. It will help y
...
+  +       .field("NESTED_FIELD2","STRING","doe")
+  +       .end() // End nested table
+       // START nested structure
+       .fields("NESTED_STRUCTURE", "NESTED_STRUCTURE_TYPE")
+           .field("latitude", "Axis", "52")
+           .field("longitude", "Axis", "13")
+           .end() // End nested structure
+       .end();
+  ```
+- Update the OData VDM to the [newest release 2102 of SAP S/4HANA Cloud](https://news.sap.com/2021/02/sap-s4hana-cloud-2102-empowering-businesses-seize-opportunities/). This includes completely new services (available as usual in packages `com.sap.cloud.sdk.s4hana.datamodel.odata.services` and `com.sap.cloud.sdk.s4hana.datamodel.odatav4.services`), new operations in previously existing services, and new entity types. The SDK supports all OData services listed in the [SAP API Business Hub for SAP S/4HANA Cloud](https://api.sap.com/shell/discover/contentpackage/SAPS4HANACloud) with the exception of the following services:
```
Comment (artemkovalyov) [inline docs/java/release-notes.mdx:144]:
```suggestion
- We updated the [OData type-safe client also known as a VDM](java/features/odata/overview.mdx) to the [newest release 2102 of SAP S/4HANA Cloud](https://news.sap.com/2021/02/sap-s4hana-cloud-2102-empowering-businesses-seize-opportunities/). This includes completely new services (available as usual in packages `com.sap.cloud.sdk.s4hana.datamodel.odata.services` and `com.sap.cloud.sdk.s4hana.datamodel.odatav4.services`), new operations in previously existing services, and new entity types. The SDK supports all OData services listed in the [SAP API Business Hub for SAP S/4HANA Cloud](https://api.sap.com/shell/discover/contentpackage/SAPS4HANACloud) with the exception of the following services:
```

File: docs/java/release-notes.mdx [hunk 13]
```diff
@@ -49,6 +49,130 @@ We highly recommend regularly updating to the latest SDK version. It will help y
...
+      - Update Resilience4J from `1.6.1` to `1.7.0`
+      - Update Java JWT Java from `3.12.0` to `3.12.1`
+      - Update PMD from `6.30.0` to `6.31.0`
+      - Update Lombok from `1.18.16` to `1.18.18`
+      - Update Jetty from `9.4.35.v20201120` to `9.4.36.v20210114`
+
+### Fixed Issues
+
+- We fixed an issue with OData V4 update requests where the serialization of primitives with types like Integer, Boolean, and Double leads to null values.
+- We fixed an issue with JCo serialization of BAPI requests, where adding field parameter values of type `ErpType` led to an exception.
```
Comment (artemkovalyov) [inline docs/java/release-notes.mdx:174]:
```suggestion
- We fixed an issue with OData V4 update requests where the serialization of primitives with types like Integer, Boolean, and Double leads to null values.
- We fixed an issue with the JCo serialization of BAPI requests, where adding field parameter values of type `ErpType` led to an exception.
```

**Review by artemkovalyov — APPROVED** [review]
It's good to go. Consider fixing the small issue I have found.

**Pull request merged**